pkg/controller/container/nodepool.go

```diff
@@ -0,0 +1,246 @@
+	if !ok {
+		return nil, errors.New(errNotNodePool)
+	}
+
+	p := &gcpv1alpha3.Provider{}
+	if err := c.kube.Get(ctx, meta.NamespacedNameOf(i.Spec.ProviderReference), p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.Secret.Namespace, Name: p.Spec.Secret.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	client, err := c.newServiceFn(ctx,
+		option.WithCredentialsJSON(s.Data[p.Spec.Secret.Key]),
+		option.WithScopes(container.CloudPlatformScope))
+	return &nodePoolExternal{cluster: client, projectID: p.Spec.ProjectID, kube: c.kube}, errors.Wrap(err, errNewClient)
+}
+
+type nodePoolExternal struct {
+	kube      client.Client
+	cluster   *container.Service
+	projectID string
+}
+
+func (e *nodePoolExternal) Observe(ctx context.Context, mg resource.Managed) (resource.ExternalObservation, error) { // nolint:gocyclo
+	cr, ok := mg.(*v1beta1.NodePool)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotNodePool)
+	}
+
+	existing, err := e.cluster.Projects.Locations.Clusters.NodePools.Get(np.GetFullyQualifiedName(cr.Spec.ForProvider, meta.GetExternalName(cr))).Context(ctx).Do()
+	if err != nil {
+		return resource.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetNodePool)
+	}
+
+	cr.Status.AtProvider = np.GenerateObservation(*existing)
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	np.LateInitializeSpec(&cr.Spec.ForProvider, *existing)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return resource.ExternalObservation{}, errors.Wrap(err, errManagedNodePoolUpdateFailed)
+		}
+	}
+
+	switch cr.Status.AtProvider.Status {
+	case v1beta1.NodePoolStateRunning:
+		cr.Status.SetConditions(v1alpha1.Available())
+		resource.SetBindable(cr)
+	case v1beta1.NodePoolStateProvisioning:
+		cr.Status.SetConditions(v1alpha1.Creating())
+	case v1beta1.NodePoolStateUnspecified, v1beta1.NodePoolStateRunningError, v1beta1.NodePoolStateError, v1beta1.NodePoolStateReconciling:
+		cr.Status.SetConditions(v1alpha1.Unavailable())
+	}
+
+	u, _ := np.IsUpToDate(&cr.Spec.ForProvider, *existing)
+
+	return resource.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: u,
+	}, nil
+}
+
+func (e *nodePoolExternal) Create(ctx context.Context, mg resource.Managed) (resource.ExternalCreation, error) {
+	cr, ok := mg.(*v1beta1.NodePool)
+	if !ok {
+		return resource.ExternalCreation{}, errors.New(errNotNodePool)
+	}
+	cr.SetConditions(v1alpha1.Creating())
+
+	// Generate GKE node pool from resource spec.
+	pool := np.GenerateNodePool(cr.Spec.ForProvider, meta.GetExternalName(cr))
+
+	create := &container.CreateNodePoolRequest{
+		NodePool: pool,
+	}
+
+	if _, err := e.cluster.Projects.Locations.Clusters.NodePools.Create(gcp.StringValue(cr.Spec.ForProvider.Cluster), create).Context(ctx).Do(); err != nil {
```

Comment:
The URI we are pulling off the `cluster` is in the form of `projects/<projectID>/locations/<clusterLocation>/clusters/<clusterName>`. This is the full path needed for creation of a `NodePool`. However, we could handle this two ways:
1. Pull the `/locations/<clusterLocation>/clusters/<clusterName` from the cluster and append the `projectID` from the `Provider`. I do not think this is a good idea because if you have clusters with the same name in the same region but in different projects then you could explicitly provider a `cluster` in the `NodePool` spec that included a `projectID` and the `NodePool` might be assigned to a cluster in a different project if it uses a different `Provider`.
1. Do what we do here where we use the full path of the `cluster` and will fail if our `Provider` doesn't have access to that project. This means that if you provide `cluster` instead of `clusterRef` you do have to have the full path including `projectID`. It is a little more verbose, but I think it is more clear when look at CRs.

Reply:
I agree that option 2 makes more sense.

> Do what we do here where we use the full path of the cluster and will fail if our Provider doesn't have access to that project

As long as we are able to document the format in an easy way and the error is clear, I think this is fine.